libc/socket/unix-socket.c

```diff
@@ -904,6 +928,587 @@ TEST(test_unix_socket, poll)
+		TEST_ASSERT_EQUAL_INT(1, poll(fds, 1, 250));
+		TEST_ASSERT_EQUAL_INT(POLLIN, fds[0].revents);
+
+		TEST_ASSERT_EQUAL_INT(0, waitpid(pid, NULL, WNOHANG));
+
+		READ_MSG(fds[0].fd, pid, 0);
+
+		fds[0].events = POLLOUT;
+		TEST_ASSERT_EQUAL_INT(1, poll(fds, 1, 250));
+		TEST_ASSERT_EQUAL_INT(POLLOUT, fds[0].revents);
+
+		SEND_MSG(fds[0].fd, pid, 0);
+
+		if (waitpid(pid, &status, 0) < 0)
+			FAIL("waitpid");
+
+		TEST_ASSERT(WIFEXITED(status));
+		TEST_ASSERT_EQUAL_INT(0, WEXITSTATUS(status));
+
+		close(fds[0].fd);
+	}
+	else {
+		if ((named = unix_named_socket(type, socket_name)) < 0)
+			exit(1);
+
+		if (listen(named, 0) < 0)
+			exit(1);
+
+		usleep(500); /* sleep so that connect would block */
+
+		if ((conn = accept(named, NULL, NULL)) < 0)
+			exit(1);
+
+		SEND_MSG(conn, pid, 0);
+
+		/* read something from parent so that parent does the first
+		 * POLLOUT before conn gets closed  */
+		READ_MSG(conn, pid, 0);
+		close(conn);
+
+		close(named);
+		unlink(socket_name);
+
+		exit(0);
+	}
+}
+
+
+static void unix_accept_connect_liveness(int type)
+{
+	unsigned int i = 0;
+
+	for (i = 0; i < 25; ++i) {
+		unix_accept_connect_liveness_helper(type);
+	}
+}
+
+
+TEST(test_unix_socket, accept_connect_liveness)
+{
+	unix_accept_connect_liveness(SOCK_STREAM);
+	unix_accept_connect_liveness(SOCK_SEQPACKET);
+}
+
+
+static void unix_socket_recv_msg_peek(int flags)
+{
+	int fd[2];
+	int msg_len = 128;
+	ssize_t n;
+
+	if (socketpair(AF_UNIX, SOCK_STREAM, 0, fd) < 0)
+		FAIL("socketpair");
+
+	n = write(fd[0], data, msg_len);
+	TEST_ASSERT(n == msg_len);
+
+	n = write(fd[0], data, 1);
+	TEST_ASSERT(n == 1);
+
+	/** Peek on first 2 iterations, on 3th do a normal read */
+	for (int i = 0; i < 4; i++) {
+		n = recv(fd[1], buf, msg_len, flags | (i < 2 ? MSG_PEEK : 0));
+		if (i < 3) {
+			/* Should read the same message 3 times */
+			TEST_ASSERT(n == msg_len);
+			TEST_ASSERT(strncmp(buf, data, msg_len) == 0);
+		}
+		else {
+			/* Should read one byte as previous message was normally read on 3rd
+			 * iteration */
+			TEST_ASSERT(n == 1);
+		}
+	}
+
+	close(fd[0]);
+	close(fd[1]);
+}
+
+
+TEST(test_unix_socket, recv_msg_peek)
+{
+	unix_socket_recv_msg_peek(0);
+	unix_socket_recv_msg_peek(MSG_DONTWAIT);
+}
+
+
+static void unix_select(int type)
+{
+	fd_set rfds;
+	struct timeval tv;
+	struct timespec ts[2];
+	int rv, ms, fd, i;
+	int bad_usecs[2] = { -1, 1000 * 000 /* 1000ms */ };
+
+	fd = socket(AF_UNIX, type, 0);
+
+	tv.tv_sec = 0;
+	tv.tv_usec = 300 * 1000;
+	FD_ZERO(&rfds);
+	FD_SET(fd, &rfds);
+	clock_gettime(CLOCK_REALTIME, &ts[0]);
+	rv = select(fd + 1, &rfds, NULL, NULL, &tv);
+	clock_gettime(CLOCK_REALTIME, &ts[1]);
+	ms = MS_BETWEEN(ts[0], ts[1]);
+	TEST_ASSERT(rv == 0);
+	TEST_ASSERT_EQUAL_INT(0, FD_ISSET(fd, &rfds));
+	TEST_ASSERT_LESS_THAN(350, ms);
+	TEST_ASSERT_GREATER_THAN(290, ms);
+
+	tv.tv_sec = 0;
+	tv.tv_usec = 1000;
+	FD_ZERO(&rfds);
+	FD_SET(fd, &rfds);
+	errno = 0;
+	rv = select(-1, &rfds, NULL, NULL, &tv);
```

Comment:
These are general tests to check if `select` correctly identifies invalid parameters. We should create a new `poll` directory and relocate the tests there, as they apply to polling various types of file descriptors.